File: app/src/main/java/com/castelijns/mmdemo/user_detail/UserDetailActivity.java
package com.castelijns.mmdemo.user_detail;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.ActionBar;
import android.view.MenuItem;
import android.widget.TextView;

import com.castelijns.mmdemo.R;
import com.castelijns.mmdemo.app.BaseActivity;
import com.castelijns.mmdemo.models.User;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.castelijns.mmdemo.users.UsersFragment.EXTRA_EMAIL_TRANSITION;
import static com.castelijns.mmdemo.users.UsersFragment.EXTRA_USER;
import static com.castelijns.mmdemo.users.UsersFragment.EXTRA_USERNAME_TRANSITION;

public class UserDetailActivity extends BaseActivity implements UserDetailContract.View {

    @BindView(R.id.tv_username)
    TextView tvUsername;

    @BindView(R.id.tv_email)
    TextView tvEmail;

    @BindView(R.id.tv_phone)
    TextView tvPhone;

    @BindView(R.id.tv_website)
    TextView tvWebsite;

    @BindView(R.id.tv_city)
    TextView tvCity;

    @BindView(R.id.tv_address)
    TextView tvAddress;

    @BindView(R.id.tv_zipcode)
    TextView tvZipcode;

    @BindView(R.id.tv_company_name)
    TextView tvCompanyName;

    @BindView(R.id.tv_company_catchphrase)
    TextView tvCompanyCatchphrase;

    @BindView(R.id.tv_company_bs)
    TextView tvCompanyBusinessStatement;

    private UserDetailContract.Presenter presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_detail);
        ButterKnife.bind(this);

        User user = getIntent().getParcelableExtra(EXTRA_USER);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setTitle(user.getName());

        // For shared element transitions.
        supportPostponeEnterTransition();

        tvUsername.setTransitionName(getIntent().getStringExtra(EXTRA_USERNAME_TRANSITION));
        tvUsername.setText(user.getUsername());

        tvEmail.setTransitionName(getIntent().getStringExtra(EXTRA_EMAIL_TRANSITION));
        tvEmail.setText(user.getEmail());

        supportStartPostponedEnterTransition();
        // end.

        tvPhone.setText(user.getPhone());
        tvWebsite.setText(user.getWebsite());

        tvCity.setText(user.getAddress().getCity());
        tvAddress.setText(String.format("%s, %s", user.getAddress().getStreet(),
                user.getAddress().getSuite()));
        tvZipcode.setText(user.getAddress().getZipcode());

        tvCompanyName.setText(user.getCompany().getName());
        tvCompanyCatchphrase.setText(user.getCompany().getCatchPhrase());
        tvCompanyBusinessStatement.setText(user.getCompany().getBs());

        presenter = new UserDetailPresenter(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/userdetail/UserDetailActivity.kt
package com.castelijns.mmdemo.userdetail

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.castelijns.mmdemo.R
import com.castelijns.mmdemo.models.User
import com.castelijns.mmdemo.users.UsersFragment.Companion.EXTRA_EMAIL_TRANSITION
import com.castelijns.mmdemo.users.UsersFragment.Companion.EXTRA_USER
import com.castelijns.mmdemo.users.UsersFragment.Companion.EXTRA_USERNAME_TRANSITION
import kotlinx.android.synthetic.main.activity_user_detail.*

class UserDetailActivity : AppCompatActivity(), UserDetailContract.View {

    private lateinit var presenter: UserDetailPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_detail)

        val user = intent.getParcelableExtra<User>(EXTRA_USER)
        presenter = UserDetailPresenter(this, user)

        val actionBar = supportActionBar
        actionBar!!.setDisplayHomeAsUpEnabled(true)
        actionBar.title = user.name

        // For shared element transitions.
        supportPostponeEnterTransition()

        tv_username.transitionName = intent.getStringExtra(EXTRA_USERNAME_TRANSITION)
        tv_username.text = user.username

        tv_email.transitionName = intent.getStringExtra(EXTRA_EMAIL_TRANSITION)
        tv_email.text = user.email

        supportStartPostponedEnterTransition()
        // end.

        tv_phone.text = user.phone
        tv_website.text = user.website

        tv_city.text = user.address.city
        tv_address.text = getString(R.string.user_detail_address, user.address.street,
                user.address.suite)
        tv_zipcode.text = user.address.zipcode

        tv_company_name.text = user.company.name
        tv_company_catchphrase.text = user.company.catchPhrase
        tv_company_bs.text = user.company.bs

        ib_directions.setOnClickListener { presenter.onDirectionsClicked() }
        presenter.start()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }

        return super.onOptionsItemSelected(item)
    }


    override fun showNavigationTo(lat: String, lon: String) {
        val intent = Intent(Intent.ACTION_VIEW,
                Uri.parse("https://www.google.com/maps/dir/?api=1&destination=$lat,$lon"))
        startActivity(intent)
    }

}
